laplace/baselaplace.py

```diff
@@ -1568,8 +1567,1025 @@ def sample(self, n_samples=100, generator=None):
+
+        As the diagonal approximation is performed with \Lambda_{MM} (which is stored in self.L),
+        the code is greatly simplified.
+        """
+        if self.diagonal_kernel:
+            self.Sigma_inv = [
+                torch.linalg.cholesky(
+                    self.gp_kernel_prior_variance * self.K_MM[c]
+                    + torch.diag(
+                        torch.nan_to_num(1.0 / (self._H_factor * lambda_c), posinf=10.0)
+                    )
+                )
+                for c, lambda_c in enumerate(self.L)
+            ]
+        else:
+            self.Sigma_inv = torch.linalg.cholesky(
+                self.gp_kernel_prior_variance * self.K_MM
+                + torch.diag(
+                    torch.nan_to_num(1 / (self._H_factor * self.L), posinf=10.0)
+                )
+            )
+
+    def _get_SoD_data_loader(self, train_loader: DataLoader) -> DataLoader:
+        """
+        Subset-of-Datapoints data loader
+        """
+        return DataLoader(
+            dataset=train_loader.dataset,
+            batch_size=train_loader.batch_size,
+            sampler=SoDSampler(N=len(train_loader.dataset), M=self.M, seed=self.seed),
+            shuffle=False,
+        )
+
+    def fit(self, train_loader, progress_bar=False):
+        """Fit the Laplace approximation of a GP posterior.
+
+        Parameters
+        ----------
+        train_loader : torch.data.utils.DataLoader
+            `train_loader.dataset` needs to be set to access \\(N\\), size of the data set
+            `train_loader.batch_size` needs to be set to access \\(b\\) batch_size
+        """
+
+        X, _ = next(iter(train_loader))
+        with torch.no_grad():
+            self.n_outputs = self.model(X[:1].to(self._device)).shape[-1]
+        setattr(self.model, 'output_size', self.n_outputs)
+        self.batch_size = train_loader.batch_size
+
+        if (
+            self.likelihood == 'regression'
+            and self.n_outputs > 1
+            and self.diagonal_kernel
+        ):
+            warnings.warn(
+                'Using FunctionalLaplace with the diagonal approximation of a GP kernel is not recommended '
+                'in the case of multivariate regression. Predictive variance will likely be overestimated.'
+            )
+
+        self.model.eval()
+
+        N = len(train_loader.dataset)
+        self.n_data = N
+
+        # by default, all training data points are used for GP inference
+        if self.M is None:
+            self.M = N
+
+        train_loader = self._get_SoD_data_loader(train_loader)
+        self.train_loader = train_loader
+        self._prior_factor_sod = self.M / self.n_data
+
+        self._init_K_MM()
+        self._init_Sigma_inv()
+
+        f, lambdas, mu = [], [], []
+
+        if progress_bar:
+            loader = enumerate(tqdm.tqdm(train_loader, desc='Fitting'))
+        else:
+            loader = enumerate(train_loader)
+
+        for i, (X, y) in loader:
```

Comment:
Do you see this `FunctionalLaplace` applicable (read: fast enough, thus useful) to foundation models like LLMs? If so, might consider allowing for custom `MutableMapping` for `X`.

See for example: https://github.com/aleximmer/Laplace/blob/main/examples/huggingface_example.md
Another example for preference learning / reward modeling: https://github.com/aleximmer/Laplace/pull/167

Reply:
Probably not useful in practice but I have adapted the code to allow it